The task solving

DanaKhromenko opened the pull request:
Please, make your review as informative as possible. I'm pretty sure I didn't understand everything about <T> and PECS and my solution was oriented on MA's tests mostly.

DanaKhromenko commented:
- removed an extra separation in import group;
- removed generic whitespaces.

RomanDubovskyi commented on src/main/java/core/mate/academy/model/Bulldozer.java:
```diff
@@ -12,4 +14,33 @@ public Bulldozer() {
     }
+
+    @Override
+    public int hashCode() {
```
ці методи доречі не обов'язково перевизначати якщо в завданні вони не використовуються

RomanDubovskyi commented on src/main/java/core/mate/academy/service/BulldozerProducer.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class BulldozerProducer implements MachineProducer {
```
```suggestion
public class BulldozerProducer implements MachineProducer<Bulldozer> {
```
тобі потрібно тут параметризувати інтерфейс конкретним типом, інакше ти можеш проставити в методі get List<String> , List<Cat> і тд, а якщо при декларації класу параметризуєш його, то воно має сваритись, якщо обрати невалдіний тип для параметризації

RomanDubovskyi commented on src/main/java/core/mate/academy/service/ExcavatorProducer.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ExcavatorProducer implements MachineProducer {
```
аналогічно тут і у інших класах

RomanDubovskyi commented on src/main/java/core/mate/academy/service/MachineProducer.java:
```diff
@@ -1,11 +1,12 @@
+import core.mate.academy.model.Machine;
 import java.util.List;
 
 public interface MachineProducer {
```
```suggestion
public interface MachineProducer<T extends Machine> {
```

RomanDubovskyi commented on src/main/java/core/mate/academy/service/MachineProducer.java:
```diff
@@ -1,11 +1,12 @@
      * @return - the list of models
      */
-    List<Object> get();
+    List<? extends Machine> get();
```
```suggestion
    List<T> get();
```
так ти встановлюєй обмеження на використовуваний тип

RomanDubovskyi commented on src/main/java/core/mate/academy/service/MachineService.java:
```diff
@@ -16,7 +17,7 @@
      * @return the list of machines
      */
-    List<Object> getAll(Class type);
+    List<T> getAll(Class type);
```
```suggestion
     List<T> getAll(Class<? extends T> type);
```

RomanDubovskyi commented on src/main/java/core/mate/academy/service/MachineServiceImpl.java:
```diff
@@ -1,7 +1,43 @@
  * Your implementation of MachineService.
  */
-public class MachineServiceImpl {
+public class MachineServiceImpl<T extends Machine> implements MachineService {
```
```suggestion
public class MachineServiceImpl implements MachineService<Machine> {
```
тут вже потірбно конкретний типом параметризовувати

RomanDubovskyi commented on src/main/java/core/mate/academy/service/MachineServiceImpl.java:
```diff
@@ -1,7 +1,43 @@
-public class MachineServiceImpl {
+public class MachineServiceImpl<T extends Machine> implements MachineService {
+    @Override
+    public List getAll(Class type) {
```
цей метод теж повинен бути параметризованим

DanaKhromenko commented:
- removed override methods: toString, hashCode and equals;
- removed unused imports;
- added diamond signature to the MachineProducer interface and it's implementations;
- replaced <T extends Machine> in get() method with <T> because of previous adding <T extends Machine> to interface signature;
- replaced input value's type Class with Class<? extends T> for making an upper bound;
- added <Machine> to the MachineService interface's call;
- replaced <T extends Machine> for class MachineServiceImpl;
- added again <T extends Machine> for class MachineServiceImpl :-)

RomanDubovskyi commented on src/main/java/core/mate/academy/service/MachineServiceImpl.java:
```diff
@@ -1,7 +1,43 @@
-public class MachineServiceImpl {
+public class MachineServiceImpl<T extends Machine> implements MachineService<Machine> {
+    @Override
+    public List getAll(Class type) {
```
```suggestion
    public List<Machine> getAll(Class<? extends Machine> type) {
```

RomanDubovskyi commented on src/main/java/core/mate/academy/service/MachineServiceImpl.java:
```diff
@@ -1,7 +1,43 @@
+    }
+
+    @Override
+    public void fill(List machines, Machine value) {
```
```suggestion
    public void fill(List<....> machines, Machine value) {
```
аналогічно в startWorking
